clippy_lints/src/non_scalar_const.rs

```diff
@@ -0,0 +1,89 @@
+use crate::rustc_target::abi::LayoutOf;
+use crate::utils::{snippet_opt, span_lint_and_then};
+use if_chain::if_chain;
+use rustc::mir::interpret::ConstValue;
+use rustc::ty::{self, ConstKind};
+use rustc_hir::{Item, ItemKind};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::InnerSpan;
+use rustc_typeck::hir_ty_to_ty;
+
+declare_clippy_lint! {
+    /// **What it does:** Checks for large `const` non-scalar types (ie: array) that should
+    /// be defined as `static` instead.
+    ///
+    /// **Why is this bad?** Performance: const variables are inlined upon use.
+    /// Static items result in only one instance and has a fixed location in memory.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    /// ```rust,ignore
+    /// // Bad code
+    /// pub const a = [0u32; 1_000_000];
+    ///
+    /// // Good code
+    /// pub static a = [0u32; 1_000_000];
+    /// ```
+    pub NON_SCALAR_CONST,
+    perf,
+    "large non-scalar const variable may cause performance overhead"
+}
+
+pub struct NonScalarConst {
+    maximum_allowed_size: u64,
+}
+
+impl NonScalarConst {
+    #[must_use]
+    pub fn new(maximum_allowed_size: u64) -> Self {
+        Self { maximum_allowed_size }
+    }
+}
+
+impl_lint_pass!(NonScalarConst => [NON_SCALAR_CONST]);
+
+impl<'a, 'tcx> LateLintPass<'a, 'tcx> for NonScalarConst {
+    fn check_item(&mut self, cx: &LateContext<'a, 'tcx>, it: &'tcx Item<'_>) {
+        // No suggestion in macros
+        if it.span.from_expansion() {
+            return;
+        }
+        if_chain! {
+            if let ItemKind::Const(hir_ty, _) = &it.kind;
+            let ty = hir_ty_to_ty(cx.tcx, hir_ty);
+            if let ty::Array(element_type, cst) = ty.kind;
+            if let ConstKind::Value(val) = cst.val;
+            if let ConstValue::Scalar(element_count) = val;
+            if let Ok(element_count) = element_count.to_machine_usize(&cx.tcx);
+            if let Ok(element_size) = cx.layout_of(element_type).map(|l| l.size.bytes());
+            if self.maximum_allowed_size < element_count * element_size;
+            then {
+                // We need to take care about the right "const" position inside the span
+                let mut start_sugg = 0;
+                if let Some(visi) = snippet_opt(cx, it.vis.span) {
+                    let len_visi = visi.len();
+                    if len_visi > 0 {
+                        start_sugg = len_visi + 1;
+                    }
+                }
+                let sugg_span = it.span.from_inner(
+                    InnerSpan::new(start_sugg, start_sugg + "const".len())
+                );
```

Comment:
Looking at the [grammar]() of const items, I would construct the span for this with the `item.ident.span.lo()`
```grammar
Visibility? const ( IDENTIFIER | _ ) : Type = Expression ;
```

Starting from `vis+1` is fine, but I would build the span like this:
```suggestion
                let sugg_span = Span::new(
                    BytePos::from_usize(start_sugg),
                    item.ident.span.lo() - BytePos::from_usize(1),
                    item.span.ctxt(),
                );
```

That way, also 

```rust
  pub  const _: u32 = 42;
//   ^^ more than one space
```

will have the correct formatting/suggestion